### PR TITLE
Configurable max gRPC message size

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Add config option for specifying max gRPC message size

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/2148

## Follow-up issue
_NA_

### CODE REVIEW

**Review by EngHabu — DISMISSED** [review]

**Comment by katrogan** [comment]
sorry, PTAL @EngHabu

**Comment by katrogan** [comment]
PTAL @EngHabu

**Comment by hamersaw** [comment]
Quick comment. I did some testing with this when I was debugging extremely large dynamic workflows (sending the workflow closure over gRPC). Once I increased the gRPC size limit to a large value (I think like 8M? I think its 4M default?!?!) I started running into timeout issues. Not sure if the plan is to scale that high, but it might be useful to add timeout configuration while this is fresh.

**Review by EngHabu — COMMENTED** [review]
+1 to Dan's comment though

File: pkg/config/config.go [hunk 1]
```diff
@@ -13,16 +13,22 @@ const SectionKey = "server"
...
 	HTTPPort             int                   `json:"httpPort" pflag:",On which http port to serve admin"`
-	GrpcPort             int                   `json:"grpcPort" pflag:",On which grpc port to serve admin"`
-	GrpcServerReflection bool                  `json:"grpcServerReflection" pflag:",Enable GRPC Server Reflection"`
+	GrpcPort             int                   `json:"grpcPort" pflag:",deprecated"`
+	GrpcServerReflection bool                  `json:"grpcServerReflection" pflag:",deprecated"`
 	KubeConfig           string                `json:"kube-config" pflag:",Path to kubernetes client config file, default is empty, useful for incluster config."`
 	Master               string                `json:"master" pflag:",The address of the Kubernetes API server."`
 	Security             ServerSecurityOptions `json:"security"`
-
+	GrpcConfig           GrpcConfig            `json:"grpcConfig"`
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/config/config.go]: done

File: pkg/config/config.go [hunk 2]
```diff
@@ -13,16 +13,22 @@ const SectionKey = "server"
...
 	// Deprecated: please use auth.AppAuth.ThirdPartyConfig instead.
 	DeprecatedThirdPartyConfig authConfig.ThirdPartyConfigOptions `json:"thirdPartyConfig" pflag:",Deprecated please use auth.appAuth.thirdPartyConfig instead."`
 }
 
+type GrpcConfig struct {
+	Port                    int  `json:"grpcPort" pflag:",On which grpc port to serve admin"`
+	ServerReflection        bool `json:"grpcServerReflection" pflag:",Enable GRPC Server Reflection"`
+	MaxGrpcMessageSizeBytes int  `json:"maxGrpcMessageSizeBytes" pflag:",The max size in bytes for incoming gRPC messages"`
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/config/config.go]: done

**Comment by katrogan** [comment]
thanks @hamersaw for the suggestion! did you see timeouts on admin or client-side? from a quick google search it seems that the grpc timeout options are only available client-side

**Comment by hamersaw** [comment]
> thanks @hamersaw for the suggestion! did you see timeouts on admin or client-side? from a quick google search it seems that the grpc timeout options are only available client-side

I'm sure you're right! It was probably client-side timeouts then (in my example set within propeller) and probably doesn't need to be addressed here.

**Comment by katrogan** [comment]
ping @EngHabu

File: pkg/config/config.go [hunk 3]
```diff
@@ -13,16 +13,22 @@ const SectionKey = "server"
...
-
+	GrpcConfig           GrpcConfig            `json:"grpc"`
 	// Deprecated: please use auth.AppAuth.ThirdPartyConfig instead.
 	DeprecatedThirdPartyConfig authConfig.ThirdPartyConfigOptions `json:"thirdPartyConfig" pflag:",Deprecated please use auth.appAuth.thirdPartyConfig instead."`
 }
 
+type GrpcConfig struct {
+	Port                    int  `json:"port" pflag:",On which grpc port to serve admin"`
+	ServerReflection        bool `json:"serverReflection" pflag:",Enable GRPC Server Reflection"`
+	MaxGrpcMessageSizeBytes int  `json:"maxMessageSizeBytes" pflag:",The max size in bytes for incoming gRPC messages"`
```
Comment (EngHabu) [inline pkg/config/config.go:29]: Nit, drop the redundant "Grpc"
```suggestion
	MaxMessageSizeBytes int  `json:"maxMessageSizeBytes" pflag:",The max size in bytes for incoming gRPC messages"`
```

**Review by EngHabu — APPROVED** [review]

File: pkg/config/config.go (hunk 3 above)
Comment (katrogan) [reply pkg/config/config.go]: d'oh thanks! this is what i get for prefix matching

**Comment by katrogan** [comment]
PTAL @EngHabu (hopefully last time 🤞)

**Review by EngHabu — APPROVED** [review]

**Pull request merged**